testing/endtoend/e2e_crash_test.go

```diff
@@ -0,0 +1,303 @@
+					return rc.GetAssertion(&bind.CallOpts{Context: neutralCtx}, creationInfo.ParentAssertionHash.Hash)
+				})
+				require.NoError(t, err)
+
+				tx, _, err := bk.Client().TransactionByHash(neutralCtx, creationInfo.TransactionHash)
+				require.NoError(t, err)
+				sender, err := gethtypes.Sender(gethtypes.NewCancunSigner(chainId), tx)
+				require.NoError(t, err)
+				honestConfirmed := sender == txOpts.From
+
+				isChallengeChild := parent.FirstChildBlock > 0 && parent.SecondChildBlock > 0
+				if !isChallengeChild {
+					// Assertion must be be a challenge child.
+					continue
+				}
+				// We expect the honest party to have confirmed it.
+				if !honestConfirmed {
+					t.Fatal("Evil party confirmed the assertion by challenge win")
+				}
+				confirmed = true
+				break
+			}
+			time.Sleep(500 * time.Millisecond) // Don't spam the backend.
+		}
+		// Once the honest, claimed assertion in the challenge is confirmed by time, we
+		// then continue the test.
+		t.Log("Assertion was confirmed by time")
+	})
+	// This test ensures that an honest validator can crash after a challenge has completed, can resync
+	// the completed challenge and continue playing the game until all essential edges are confirmed.
+	// This is to ensure that even if a challenge is completed, we can still resync it and continue
+	// playing for the sake of refunding honest stakes.
+	t.Run(
+		"crashes once challenged assertion is confirmed and restarts to confirm essential edges",
+		func(t *testing.T) {
+			// We start by canceling the honest context, and then wait a few seconds, and expect the honest
+			// validator to resync the challenge that just completed to continue playing the game
+			// until all essential root edges are confirmed by time.
+			honestCancel()
+			t.Log("Honest context has been canceled after challenge win")
+
+			// We then restart the honest validator after a few seconds of wait time.
+			time.Sleep(time.Second * 30)
+
+			honestCtx, honestCancel = context.WithCancel(context.Background())
+			honestChain = setupAssertionChain(t, honestCtx, bk.Client(), rollupAddr.Rollup, txOpts)
+			honestManager, err = cm.NewChallengeStack(honestChain, honestStateManager, honestOpts...)
+			require.NoError(t, err)
+
+			honestManager.Start(honestCtx)
+
+			t.Log("Restarted honest validator to continue playing game after challenge has finished")
+
+			// We then expect that all essential root edges created by the honest validator are confirmed by time.
+			// Scrape all the honest edges onchain (the ones made by the honest address).
+			// Check if the edges that have claim id != None are confirmed (those are essential root edges)
+			// and also check one step edges from honest party are confirmed.
+			honestEssentialRootIds := make(map[common.Hash]bool, 0)
+			chainId, err := bk.Client().ChainID(neutralCtx)
+			require.NoError(t, err)
+			it, err := cmBindings.FilterEdgeAdded(nil, nil, nil, nil)
+			require.NoError(t, err)
+			for it.Next() {
```

Comment:
Why doesn't this one have a similar pattern to above where we have a loop as long as the neutral context is alive and then keep working through the filterEdgeAdded iterator until we find everything we're looking for with a half-second delay to keep from spamming the backend?

Reply:
We don't know how many essential edges we will have, as it depends on the test. I guess it could be computed from the test params, but given it passes as is, I'd rather keep it simple for now